File: archive/leetcode-go/internal/code/001.go
package code

func twoSum(nums []int, target int) []int {
	record := map[int]int{}
	for p, v := range nums {
		remain := target - v
		if i, ok := record[remain]; ok {
			return []int{i, p}
		} else {
			record[v] = p
		}
	}
	return nil
}

File: leetcode/001.cpp
#include <cstddef>
#include <iostream>
#include <iterator>
#include <unordered_map>

#include "../template/leetcode.h"
#include "../template/utils.h"

using namespace std;

class Solution {
 public:
  vector<int> twoSum(vector<int>& nums, int target) {
    std::unordered_map<int, int> store;
    for (size_t i = 0ul; i < nums.size(); ++i) {
      auto& num = nums[i];
      auto idx = store[num];
      if (idx > 0) {
        return {idx - 1, int(i)};
      } else {
        store[target - num] = i + 1;
      }
    }
    return {};
  }
};

int main() {
  // mock data
  std::vector<int> v = {2, 7, 11, 15};
  int target = 9;
  Solution s;
  printVector(s.twoSum(v, target));
  return 0;
}
